File: gps-tracker-app/android/app/src/main/java/com/nicotomassini/gpstracker/clustering/NativeMapClusteringModule.kt
package com.nicotomassini.gpstracker.clustering

import com.facebook.react.bridge.*
import java.util.concurrent.Executors
import kotlin.math.floor
import kotlin.math.pow

private data class ClusterEntity(
  val id: String,
  val latitude: Double,
  val longitude: Double,
  val count: Int,
  val leafPointIndices: IntArray,
  val isCluster: Boolean,
)

private data class ZoomSnapshot(
  val zoom: Int,
  val entities: List<ClusterEntity>,
)

private data class DatasetCacheEntry(
  val key: String,
  val points: ReadableArray,
  val snapshots: Map<Int, ZoomSnapshot>,
  val leaves: Map<String, IntArray>,
  val expansionZoom: Map<String, Int>,
)

class NativeMapClusteringModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {
  private val worker = Executors.newSingleThreadExecutor()
  private val cache = LinkedHashMap<String, DatasetCacheEntry>(16, 0.75f, true)
  private val maxCacheEntries = 8
  private var lastLeaves: Map<String, IntArray> = emptyMap()
  private var lastPoints: ReadableArray? = null
  private var lastExpansionZoom: Map<String, Int> = emptyMap()

  override fun getName(): String = "NativeMapClusteringModule"

  @ReactMethod
  fun buildClusters(
    points: ReadableArray,
    zoom: Double,
    bounds: ReadableMap,
    options: ReadableMap?,
    promise: Promise
  ) {
    worker.execute {
      try {
        val datasetId = if (options?.hasKey("datasetId") == true) options.getString("datasetId") else "default"
        val radius = if (options?.hasKey("radius") == true) options.getDouble("radius") else 56.0
        val minPoints = if (options?.hasKey("minPoints") == true) options.getInt("minPoints") else 3
        val maxZoom = if (options?.hasKey("maxZoom") == true) options.getInt("maxZoom") else 18
        val minZoom = if (options?.hasKey("minZoom") == true) options.getInt("minZoom") else 0
        val z = zoom.toInt().coerceIn(minZoom, maxZoom)

        // Spatial Filtering (Culling)
        val north = if (bounds.hasKey("north")) bounds.getDouble("north") else 90.0
        val south = if (bounds.hasKey("south")) bounds.getDouble("south") else -90.0
        val east = if (bounds.hasKey("east")) bounds.getDouble("east") else 180.0
        val west = if (bounds.hasKey("west")) bounds.getDouble("west") else -180.0
        
        val latPad = Math.abs(north - south) * 0.25
        val lonPad = Math.abs(east - west) * 0.25
        
        val bNorth = (north + latPad).coerceAtMost(90.0)
        val bSouth = (south - latPad).coerceAtLeast(-90.0)
        val bEast = east + lonPad
        val bWest = west - lonPad
        
        val filteredIndices = mutableListOf<Int>()
        for (i in 0 until points.size()) {
          val p = points.getMap(i) ?: continue
          val lat = p.getDouble("latitude")
          val lon = p.getDouble("longitude")
          
          if (lat in bSouth..bNorth) {
            if (bWest <= bEast) {
              if (lon in bWest..bEast) filteredIndices.add(i)
            } else {
              if (lon >= bWest || lon <= bEast) filteredIndices.add(i)
            }
          }
        }

        val cacheKey = "${datasetId}|${filteredIndices.size}|$radius|$minPoints|$maxZoom|$minZoom"

        val cached = cache[cacheKey]
        if (cached != null) {
          // Movement detection
          if (filteredIndices.isNotEmpty()) {
            val f = filteredIndices.first()
            val l = filteredIndices.last()
            val p1 = points.getMap(f)
            val p2 = points.getMap(l)
            val cp1 = cached.points.getMap(f)
            val cp2 = cached.points.getMap(l)
            if (p1?.getDouble("latitude") == cp1?.getDouble("latitude") &&
                p2?.getDouble("latitude") == cp2?.getDouble("latitude")) {
              val snapshot = cached.snapshots[z] ?: ZoomSnapshot(z, emptyList())
              val out = toOutput(snapshot)
              lastLeaves = cached.leaves
              lastPoints = cached.points
              lastExpansionZoom = cached.expansionZoom
              reactApplicationContext.runOnUiQueueThread { promise.resolve(out) }
              return@execute
            }
          }
        }

        val built = buildHierarchy(points, filteredIndices, radius, minPoints, maxZoom, minZoom)
        val snapshot = built.first[z] ?: ZoomSnapshot(z, emptyList())
        val out = toOutput(snapshot)
        lastLeaves = built.second
        lastPoints = points
        lastExpansionZoom = built.third

        putCache(
          cacheKey,
          DatasetCacheEntry(
            key = cacheKey,
            points = points,
            snapshots = built.first,
            leaves = built.second,
            expansionZoom = built.third,
          )
        )
        reactApplicationContext.runOnUiQueueThread { promise.resolve(out) }
      } catch (t: Throwable) {
        reactApplicationContext.runOnUiQueueThread { promise.reject("clustering_error", t.message, t) }
      }
    }
  }

  @ReactMethod
  fun getLeaves(clusterId: String, limit: Int, offset: Int, promise: Promise) {
    val pointIndices = lastLeaves[clusterId] ?: run {
      promise.resolve(Arguments.createArray())
      return
    }
    val src = lastPoints ?: run {
      promise.resolve(Arguments.createArray())
      return
    }
    val start = offset.coerceAtLeast(0)
    val lim = limit.coerceAtLeast(0)
    if (lim == 0 || start >= pointIndices.size) {
      promise.resolve(Arguments.createArray())
      return
    }
    val end = (start + lim).coerceAtMost(pointIndices.size)
    val out = Arguments.createArray()
    for (i in start until end) {
      val idx = pointIndices[i]
      val map = src.getMap(idx)
      if (map != null) out.pushMap(map)
    }
    promise.resolve(out)
  }

  @ReactMethod
  fun getExpansionZoom(clusterId: String, promise: Promise) {
    promise.resolve(lastExpansionZoom[clusterId] ?: 18)
  }

  private fun putCache(key: String, value: DatasetCacheEntry) {
    cache[key] = value
    while (cache.size > maxCacheEntries) {
      val first = cache.entries.firstOrNull()?.key ?: break
      cache.remove(first)
    }
  }

  private fun buildHierarchy(
    points: ReadableArray,
    filteredIndices: List<Int>,
    radius: Double,
    minPoints: Int,
    maxZoom: Int,
    minZoom: Int,
  ): Triple<Map<Int, ZoomSnapshot>, Map<String, IntArray>, Map<String, Int>> {
    var current = mutableListOf<ClusterEntity>()
    for (i in filteredIndices) {
      val p = points.getMap(i) ?: continue
      current.add(
        ClusterEntity(
          id = "p_$i",
          latitude = p.getDouble("latitude"),
          longitude = p.getDouble("longitude"),
          count = 1,
          leafPointIndices = intArrayOf(i),
          isCluster = false,
        )
      )
    }

    val snapshots = mutableMapOf<Int, ZoomSnapshot>()
    val leaves = mutableMapOf<String, IntArray>()
    val expansionMap = mutableMapOf<String, Int>()

    for (zoom in maxZoom downTo minZoom) {
      val tileScale = 2.0.pow(zoom.toDouble())
      val degPerPixelLon = 360.0 / (256.0 * tileScale)
      val zoomRadiusScale = when {
        zoom >= 15 -> 0.65
        zoom >= 12 -> 0.80
        zoom <= 5 -> 1.30
        else -> 1.00
      }
      val cellLon = (radius * zoomRadiusScale * degPerPixelLon).coerceAtLeast(0.000001)
      val cellLat = cellLon

      val buckets = mutableMapOf<String, MutableList<ClusterEntity>>()
      current.forEach { e ->
        val gx = floor((e.longitude + 180.0) / cellLon).toInt()
        val gy = floor((e.latitude + 90.0) / cellLat).toInt()
        val key = "${gx}_${gy}"
        buckets.getOrPut(key) { mutableListOf() }.add(e)
      }

      val next = mutableListOf<ClusterEntity>()
      // Optimization: Avoid sorted map
      buckets.forEach { (_, bucket) ->
        val total = bucket.sumOf { it.count }
        if (total >= minPoints) {
          val wLat = bucket.sumOf { it.latitude * it.count.toDouble() }
          val wLon = bucket.sumOf { it.longitude * it.count.toDouble() }
          
          // Optimization: Simple merge instead of distinct/sorted
          val leafs = IntArray(total)
          var pos = 0
          for (e in bucket) {
            System.arraycopy(e.leafPointIndices, 0, leafs, pos, e.leafPointIndices.size)
            pos += e.leafPointIndices.size
          }
          
          val id = "c_${zoom}_${leafs.firstOrNull() ?: -1}_${leafs.size}"
          val cluster = ClusterEntity(
            id = id,
            latitude = wLat / total.toDouble(),
            longitude = wLon / total.toDouble(),
            count = total,
            leafPointIndices = leafs,
            isCluster = true,
          )
          next.add(cluster)
          leaves[id] = leafs
          expansionMap[id] = (zoom + 1).coerceAtMost(maxZoom)
        } else {
          next.addAll(bucket)
        }
      }
      snapshots[zoom] = ZoomSnapshot(zoom, next.toList())
      current = next
    }

    return Triple(snapshots, leaves, expansionMap)
  }

  @ReactMethod
  fun buildFullHierarchyGeoJSON(
    points: ReadableArray,
    options: ReadableMap?,
    promise: Promise
  ) {
    worker.execute {
      try {
        val radius = if (options?.hasKey("radius") == true) options.getDouble("radius") else 56.0
        val minPoints = if (options?.hasKey("minPoints") == true) options.getInt("minPoints") else 3
        val maxZoom = if (options?.hasKey("maxZoom") == true) options.getInt("maxZoom") else 18
        val minZoom = if (options?.hasKey("minZoom") == true) options.getInt("minZoom") else 0

        val validIndices = mutableListOf<Int>()
        for (i in 0 until points.size()) {
          val p = points.getMap(i) ?: continue
          if (p.hasKey("latitude") && p.hasKey("longitude")) {
            validIndices.add(i)
          }
        }

        val built = buildHierarchy(points, validIndices, radius, minPoints, maxZoom, minZoom)
        
        val geoJsonOutput = Arguments.createMap()
        built.first.forEach { (zoom, snapshot) ->
          geoJsonOutput.putMap(zoom.toString(), toGeoJsonFeatureCollection(snapshot))
        }

        lastLeaves = built.second
        lastPoints = points
        lastExpansionZoom = built.third

        reactApplicationContext.runOnUiQueueThread { promise.resolve(geoJsonOutput) }
      } catch (t: Throwable) {
        reactApplicationContext.runOnUiQueueThread { promise.reject("clustering_error", t.message, t) }
      }
    }
  }

  private fun toGeoJsonFeatureCollection(snapshot: ZoomSnapshot): WritableMap {
    val features = Arguments.createArray()
    snapshot.entities.forEach { e ->
      val isCluster = e.isCluster && e.count > 1
      val feature = Arguments.createMap()
      feature.putString("type", "Feature")
      feature.putString("id", e.id)
      
      val geometry = Arguments.createMap()
      geometry.putString("type", "Point")
      val coords = Arguments.createArray()
      coords.pushDouble(e.longitude)
      coords.pushDouble(e.latitude)
      geometry.putArray("coordinates", coords)
      feature.putMap("geometry", geometry)
      
      val props = Arguments.createMap()
      props.putBoolean("cluster", isCluster)
      props.putInt("point_count", e.count)
      props.putString("type", if (isCluster) "cluster" else "point")
      feature.putMap("properties", props)
      
      features.pushMap(feature)
    }
    
    val collection = Arguments.createMap()
    collection.putString("type", "FeatureCollection")
    collection.putArray("features", features)
    return collection
  }

  private fun toOutput(snapshot: ZoomSnapshot): WritableArray {
    val out = Arguments.createArray()
    snapshot.entities.forEach { e ->
      val item = Arguments.createMap()
      item.putString("id", e.id)
      item.putString("type", if (e.isCluster && e.count > 1) "cluster" else "point")
      item.putInt("count", e.count)
      item.putDouble("latitude", e.latitude)
      item.putDouble("longitude", e.longitude)
      out.pushMap(item)
    }
    return out
  }
}

File: gps-tracker-app/ios/Trackly/NativeMapClusteringModule.swift
import Foundation
import React
import CoreLocation

private struct ClusterEntity {
  let id: String
  let latitude: Double
  let longitude: Double
  let count: Int
  let leafPointIndices: [Int]
  let isCluster: Bool
}

private struct ZoomSnapshot {
  let zoom: Int
  let entities: [ClusterEntity]
}

private struct DatasetCacheEntry {
  let key: String
  let createdAt: TimeInterval
  let points: [[String: Any]]
  let snapshots: [Int: ZoomSnapshot]
  let entitiesById: [String: ClusterEntity]
}

@objc(NativeMapClusteringModule)
class NativeMapClusteringModule: NSObject {
  private let workQueue = DispatchQueue(label: "clustering.native.ios.queue", qos: .userInitiated)
  private var datasetCache: [String: DatasetCacheEntry] = [:]
  private var cacheOrder: [String] = []
  private let maxCacheEntries = 8
  private var lastLeaves: [String: [[String: Any]]] = [:]
  private var lastExpansionZoom: [String: Int] = [:]

  @objc static func requiresMainQueueSetup() -> Bool { false }

  @objc(buildClusters:zoom:bounds:options:resolver:rejecter:)
  func buildClusters(
    _ points: [[String: Any]],
    zoom: NSNumber,
    bounds: [String: Any],
    options: [String: Any]?,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock
  ) {
    workQueue.async {
      do {
        let datasetId = (options?["datasetId"] as? String) ?? "default"
        let radius = (options?["radius"] as? NSNumber)?.doubleValue ?? 56.0
        let minPoints = (options?["minPoints"] as? NSNumber)?.intValue ?? 3
        let maxZoom = (options?["maxZoom"] as? NSNumber)?.intValue ?? 18
        let minZoom = (options?["minZoom"] as? NSNumber)?.intValue ?? 0
        let z = max(minZoom, min(maxZoom, zoom.intValue))

        // Spatial Filtering (Culling)
        // Add a 25% buffer to avoid popping at edges during small pans
        let north = (bounds["north"] as? NSNumber)?.doubleValue ?? 90.0
        let south = (bounds["south"] as? NSNumber)?.doubleValue ?? -90.0
        let east = (bounds["east"] as? NSNumber)?.doubleValue ?? 180.0
        let west = (bounds["west"] as? NSNumber)?.doubleValue ?? -180.0
        
        let latPad = abs(north - south) * 0.25
        let lonPad = abs(east - west) * 0.25
        
        let bNorth = min(90, north + latPad)
        let bSouth = max(-90, south - latPad)
        let bEast = east + lonPad
        let bWest = west - lonPad
        
        // Filter points within padded bounds
        var filteredIndices: [Int] = []
        filteredIndices.reserveCapacity(points.count)
        
        for (idx, p) in points.enumerated() {
          guard let lat = (p["latitude"] as? NSNumber)?.doubleValue,
                let lon = (p["longitude"] as? NSNumber)?.doubleValue else { continue }
          
          if lat >= bSouth && lat <= bNorth {
            // Handle wrap around for longitude
            if bWest <= bEast {
              if lon >= bWest && lon <= bEast { filteredIndices.append(idx) }
            } else {
              if lon >= bWest || lon <= bEast { filteredIndices.append(idx) }
            }
          }
        }

        // Cache Key should include some data characteristic if possible, 
        // or we just skip cache for very small datasets or frequent updates.
        // For now, let's keep it simple but optimized.
        let cacheKey = "\(datasetId)|\(filteredIndices.count)|\(radius)|\(minPoints)|\(maxZoom)|\(minZoom)"

        if let cached = self.datasetCache[cacheKey] {
          // Verify first and last points to detect movement if count is same
          if !filteredIndices.isEmpty {
            let firstIdx = filteredIndices[0]
            let lastIdx = filteredIndices.last!
            let p1 = points[firstIdx]
            let p2 = points[lastIdx]
            let cachedPoints = cached.points
            if cachedPoints.count > max(firstIdx, lastIdx) {
              let cp1 = cachedPoints[firstIdx]
              let cp2 = cachedPoints[lastIdx]
              if (p1["latitude"] as? Double) == (cp1["latitude"] as? Double) &&
                 (p2["latitude"] as? Double) == (cp2["latitude"] as? Double) {
                let output = self.toOutput(
                  snapshot: cached.snapshots[z] ?? ZoomSnapshot(zoom: z, entities: []),
                  points: cached.points
                )
                self.lastLeaves = cached.leaves
                self.lastExpansionZoom = cached.expansionZoom
                DispatchQueue.main.async { resolver(output) }
                return
              }
            }
          }
        }

        let built = self.buildHierarchy(
          points: points,
          filteredIndices: filteredIndices,
          radius: radius,
          minPoints: minPoints,
          maxZoom: maxZoom,
          minZoom: minZoom
        )
        
        let snapshot = built.snapshots[z] ?? ZoomSnapshot(zoom: z, entities: [])
        let output = self.toOutput(snapshot: snapshot, points: points)
        
        // Only store the maps for the CURRENT view to save memory/CPU
        // or build them on the fly in getLeaves.
        // For now, we still return the full snapshot at Z.
        self.lastLeaves = built.leaves
        self.lastExpansionZoom = built.expansionZoom
        
        self.insertCache(
          cacheKey: cacheKey,
          entry: DatasetCacheEntry(
            key: cacheKey,
            createdAt: Date().timeIntervalSince1970,
            points: points,
            snapshots: built.snapshots,
            entitiesById: built.entitiesById,
            leaves: built.leaves,
            expansionZoom: built.expansionZoom
          )
        )
        DispatchQueue.main.async { resolver(output) }
      } catch {
        DispatchQueue.main.async {
          rejecter("clustering_error", error.localizedDescription, error)
        }
      }
    }
  }

  @objc(getLeaves:limit:offset:resolver:rejecter:)
  func getLeaves(
    _ clusterId: String,
    limit: NSNumber,
    offset: NSNumber,
    resolver: RCTPromiseResolveBlock,
    rejecter: RCTPromiseRejectBlock
  ) {
    let all = lastLeaves[clusterId] ?? []
    let o = max(0, offset.intValue)
    let l = max(0, limit.intValue)
    if o >= all.count || l == 0 {
      resolver([])
      return
    }
    let end = min(all.count, o + l)
    resolver(Array(all[o..<end]))
  }

  @objc(getExpansionZoom:resolver:rejecter:)
  func getExpansionZoom(
    _ clusterId: String,
    resolver: RCTPromiseResolveBlock,
    rejecter: RCTPromiseRejectBlock
  ) {
    resolver(lastExpansionZoom[clusterId] ?? 18)
  }

  private func insertCache(cacheKey: String, entry: DatasetCacheEntry) {
    datasetCache[cacheKey] = entry
    cacheOrder.removeAll(where: { $0 == cacheKey })
    cacheOrder.append(cacheKey)
    while cacheOrder.count > maxCacheEntries {
      let evicted = cacheOrder.removeFirst()
      datasetCache.removeValue(forKey: evicted)
    }
  }

  private func buildHierarchy(
    points: [[String: Any]],
    filteredIndices: [Int],
    radius: Double,
    minPoints: Int,
    maxZoom: Int,
    minZoom: Int
  ) -> (snapshots: [Int: ZoomSnapshot], entitiesById: [String: ClusterEntity], leaves: [String: [[String: Any]]], expansionZoom: [String: Int]) {
    var entities: [ClusterEntity] = []
    entities.reserveCapacity(filteredIndices.count)

    for idx in filteredIndices {
      let point = points[idx]
      guard let lat = (point["latitude"] as? NSNumber)?.doubleValue,
            let lon = (point["longitude"] as? NSNumber)?.doubleValue else { continue }
      entities.append(
        ClusterEntity(
          id: "p_\(idx)",
          latitude: lat,
          longitude: lon,
          count: 1,
          leafPointIndices: [idx],
          isCluster: false
        )
      )
    }

    var snapshots: [Int: ZoomSnapshot] = [:]
    var entitiesById: [String: ClusterEntity] = [:]
    var leavesMap: [String: [[String: Any]]] = [:]
    var expansionMap: [String: Int] = [:]
    
    var current = entities

    for zoom in stride(from: maxZoom, through: minZoom, by: -1) {
      let tileScale = pow(2.0, Double(zoom))
      let degPerPixelLon = 360.0 / (256.0 * tileScale)
      let zoomRadiusScale: Double
      if zoom >= 15 {
        zoomRadiusScale = 0.65
      } else if zoom >= 12 {
        zoomRadiusScale = 0.80
      } else if zoom <= 5 {
        zoomRadiusScale = 1.30
      } else {
        zoomRadiusScale = 1.00
      }
      let cellLon = max(0.000001, radius * zoomRadiusScale * degPerPixelLon)
      let cellLat = cellLon

      var buckets: [String: [ClusterEntity]] = [:]
      for e in current {
        let gx = Int(floor((e.longitude + 180.0) / cellLon))
        let gy = Int(floor((e.latitude + 90.0) / cellLat))
        let key = "\(gx)_\(gy)"
        buckets[key, default: []].append(e)
      }

      var next: [ClusterEntity] = []
      next.reserveCapacity(current.count)

      // Avoid sorting buckets for performance
      for (_, bucket) in buckets {
        let total = bucket.reduce(0) { $0 + $1.count }
        if total >= minPoints {
          let weightedLat = bucket.reduce(0.0) { $0 + $1.latitude * Double($1.count) }
          let weightedLon = bucket.reduce(0.0) { $0 + $1.longitude * Double($1.count) }
          
          // Optimization: Merge leaf indices without Set/Sort
          // Since building bottom-up, indices are already unique and somewhat ordered
          var allLeafIndices: [Int] = []
          allLeafIndices.reserveCapacity(total)
          for e in bucket {
            allLeafIndices.append(contentsOf: e.leafPointIndices)
          }
          
          let firstLeaf = allLeafIndices.first ?? -1
          let id = "c_\(zoom)_\(firstLeaf)_\(allLeafIndices.count)"
          let cluster = ClusterEntity(
            id: id,
            latitude: weightedLat / Double(total),
            longitude: weightedLon / Double(total),
            count: total,
            leafPointIndices: allLeafIndices,
            isCluster: true
          )
          next.append(cluster)
          entitiesById[id] = cluster
          
          // Pre-populate maps only for clusters
          leavesMap[id] = allLeafIndices.compactMap { idx in
            if idx >= 0 && idx < points.count { return points[idx] }
            return nil
          }
          expansionMap[id] = min(maxZoom, zoom + 1)
        } else {
          for e in bucket {
            next.append(e)
            entitiesById[e.id] = e
          }
        }
      }

      snapshots[zoom] = ZoomSnapshot(zoom: zoom, entities: next)
      current = next
    }

    return (snapshots, entitiesById, leavesMap, expansionMap)
  }

  @objc(buildFullHierarchyGeoJSON:options:resolver:rejecter:)
  func buildFullHierarchyGeoJSON(
    _ points: [[String: Any]],
    options: [String: Any]?,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock
  ) {
    workQueue.async {
      do {
        let datasetId = (options?["datasetId"] as? String) ?? "default"
        let radius = (options?["radius"] as? NSNumber)?.doubleValue ?? 56.0
        let minPoints = (options?["minPoints"] as? NSNumber)?.intValue ?? 3
        let maxZoom = (options?["maxZoom"] as? NSNumber)?.intValue ?? 18
        let minZoom = (options?["minZoom"] as? NSNumber)?.intValue ?? 0

        // Spatial filtering for "Full Hierarchy" is less useful since we want all zooms,
        // but we can still cull points that are completely outside the world (safety).
        var validIndices: [Int] = []
        validIndices.reserveCapacity(points.count)
        for (idx, p) in points.enumerated() {
          if let _ = p["latitude"] as? NSNumber, let _ = p["longitude"] as? NSNumber {
            validIndices.append(idx)
          }
        }

        let built = self.buildHierarchy(
          points: points,
          filteredIndices: validIndices,
          radius: radius,
          minPoints: minPoints,
          maxZoom: maxZoom,
          minZoom: minZoom
        )

        var geoJsonOutput: [String: [String: Any]] = [:]
        for (zoom, snapshot) in built.snapshots {
          geoJsonOutput["\(zoom)"] = self.toGeoJsonFeatureCollection(snapshot: snapshot)
        }

        // Cache the maps for getLeaves/getExpansionZoom
        self.lastLeaves = built.leaves
        self.lastExpansionZoom = built.expansionZoom

        DispatchQueue.main.async { resolver(geoJsonOutput) }
      } catch {
        DispatchQueue.main.async {
          rejecter("clustering_error", error.localizedDescription, error)
        }
      }
    }
  }

  private func toGeoJsonFeatureCollection(snapshot: ZoomSnapshot) -> [String: Any] {
    var features: [[String: Any]] = []
    features.reserveCapacity(snapshot.entities.count)
    
    for e in snapshot.entities {
      let isCluster = e.isCluster && e.count > 1
      features.append([
        "type": "Feature",
        "id": e.id,
        "geometry": [
          "type": "Point",
          "coordinates": [e.longitude, e.latitude]
        ],
        "properties": [
          "cluster": isCluster,
          "point_count": e.count,
          "type": isCluster ? "cluster" : "point"
        ]
      ])
    }
    
    return [
      "type": "FeatureCollection",
      "features": features
    ]
  }

  private func toOutput(
    snapshot: ZoomSnapshot,
    points: [[String: Any]]
  ) -> [[String: Any]] {
    var out: [[String: Any]] = []
    out.reserveCapacity(snapshot.entities.count)
    for e in snapshot.entities {
      out.append([
        "id": e.id,
        "type": e.isCluster && e.count > 1 ? "cluster" : "point",
        "count": e.count,
        "latitude": e.latitude,
        "longitude": e.longitude,
      ])
    }
    return out
  }
}
